### PR TITLE
Fixes float misplace decimal issue

### LINKED ISSUE #13638: Float from string with underscores misplaces decimal point
This is correct:
````
In [52]: Float('1_234.345678', 24)
Out[52]: 1234.34567800000000000000
````

None of these are:
````
In [53]: Float('1_234.345_678', 24)
Out[53]: 123.434567800000000000000

In [54]: Float('1_234.34_5_678', 24)
Out[54]: 12.3434567800000000000000

In [55]: Float('1_234.34_5_6_78', 24)
Out[55]: 1.23434567800000000000000

In [56]: Float('1_234.34_5_6_7_8', 24)
Out[56]: 0.123434567800000000000000
````

I think this is an upstream bug in mpmath: https://github.com/fredrik-johansson/mpmath/issues/377

### PR DESCRIPTION
**Pull request by shivanikohlii**
#### References to other Issues or PRs
Fixes #13638

#### Brief description of what is fixed or changed
* Fixed the misplaced, now output
In [52]: Float('1_234.345678', 24)
Out[52]: 1234.34567800000000000000

#### Release Notes

* core
  * string input to Float with underscores past the decimal and specified precision, e.g. Float('1.234_5', 12), now parsed correctly

### CODE REVIEW

**Comment by divyanshu132** [comment]
Concatenate the issue number with `#`, `Fixes #13638` then it will be easy for the reviewer to check the actual issue.

**Comment by smichr** [comment]
Please add a test that fails in master, like
```
assert Float('1_23.4_5_6', 12) == Float('123.456', 12)
```

File: sympy/core/numbers.py [hunk 1]
```diff
@@ -1013,6 +1013,7 @@ def __new__(cls, num, dps=None, prec=None, precision=None):
         if isinstance(num, float):
             _mpf_ = mlib.from_float(num, precision, rnd)
         elif isinstance(num, string_types):
+            num = num.replace('_','')
```
Comment (smichr) [inline sympy/core/numbers.py]: please add comment like
```python
# next line is workaround for https://github.com/fredrik-johansson/mpmath/issues/377
num = num.replace('_','')
```

**Comment by smichr** [comment]
I made an edit of the release notes, otherwise with the updated test this will be good.
Just in case anything ever changes for the non-precision case, an additional test like `assert Float('1_23.4_56') == Float('123.456')` would be good to add, too.

**Comment by asmeurer** [comment]
Technically this will allow things that aren't valid float literals. There can't be two consecutive underscores, and an underscore cannot be next to a nonnumeric character in the string, or at the beginning or the end.

**Comment by asmeurer** [comment]
Actually, it would be better to fix this upstream in mpmath. See https://github.com/fredrik-johansson/mpmath/issues/442 and https://github.com/fredrik-johansson/mpmath/issues/377. 

OTOH since it gives wrong results currently, maybe a workaround guard in SymPy is OK.

**Comment by smichr** [comment]
> this will allow things that aren't valid float literals

There is a guard against this in libmpf.str_to_man_exp:
```
    # Verify that the input is a valid float literal
    float(x)
```

**Comment by smichr** [comment]
Allowing underscores in numbers is something introduced by [PEP515](https://www.python.org/dev/peps/pep-0515/) and added in Python 3.6. So what we could do is validate that the python the user is using allows this syntax and that it is correct (addressing @asmeurer's concern):

```python
    # check that float is valid if it has _ in it
    if '_' in num:
        float(num)
    # correct as workaround for mpmath not handling
    # underscore in decimal portion
    num = num.replace('_', '')
    _mpf_ = mlib.from_str(num, precision, rnd)
```

**Comment by asmeurer** [comment]
I think there may also be some code in the sympy_parser module that does this even in Python < 3.6.

**Comment by smichr** [comment]
> does this even in Python < 3.6

When I try in 2.7 I get
```python
>>> from sympy.parsing.sympy_parser import *
>>> parse_expr('1.23_45')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "sympy\parsing\sympy_parser.py", line 966, in parse_expr
    return eval_expr(code, local_dict, global_dict)
  File "sympy\parsing\sympy_parser.py", line 879, in eval_expr
    code, global_dict, local_dict)  # take local objects in preference
  File "<string>", line 1
    Float ('1.23' )Symbol ('_45' )
                        ^
SyntaxError: invalid syntax
```
But it parses correctly in 3.6. A search for underscores didn't show any special handling there.

**Comment by smichr** [comment]
@shivanikohlii , how about this:
```python
            # workaround for
            # https://github.com/fredrik-johansson/mpmath/issues/377
            if '_' in num:
                float(num)  # passes if it is a valid float in Py >= 3.6
            # older versions of mpmath do not properly handle the _ past
            # the decimal, so delete them all since we already know the
            # form is a valid Python float
            num = num.replace('_', '')
            _mpf_ = mlib.from_str(num, precision, rnd)
```

**Comment by shivanikohlii** [comment]
> ```python
> # older versions of mpmath do not properly handle the _ past
>  # the decimal, so delete them all since we already know the
> ```
Do you want me to implement something here that deletes the value if not Py>= 3.6?

**Comment by smichr** [comment]
> not Py>= 3.6?

Initially, I would say no since then the `parse_expr` would behave differently than `Float`. But `Float` is already tolerant to spaces used like digit separators. So the only change that would be needed, I think, is:
```diff
diff --git a/sympy/core/numbers.py b/sympy/core/numbers.py
index 27ca315..c2a86e0 100644
--- a/sympy/core/numbers.py
+++ b/sympy/core/numbers.py
@@ -806,10 +806,11 @@ class Float(Number):
     100.0

     Float can automatically count significant figures if a null string
-    is sent for the precision; space are also allowed in the string. (Auto-
-    counting is only allowed for strings, ints and longs).
+    is sent for the precision; spaces or underscores are also allowed
+    in the string. (Auto-counting is only allowed for strings, ints
+    and longs).

-    >>> Float('123 456 789 . 123 456', '')
+    >>> Float('123 456 789 ._123_456', '')
     123456789.123456
     >>> Float('12e-3', '')
     0.012
@@ -944,6 +945,10 @@ def __new__(cls, num, dps=None, prec=None, precision=None):

         if isinstance(num, string_types):
             num = num.replace(' ', '')
+            if num.startswith('_'):
+                raise ValueError(
+                    'underscore not allowed at the start of a number')
+            num = num.replace('_', '')
             if num.startswith('.') and len(num) > 1:
                 num = '0' + num
             elif num.startswith('-.') and len(num) > 2:
```

A test, too.

**Comment by shivanikohlii** [comment]
> A test, too.

I made two tests in test_numbers.py. I'm not a 100% sure if they are in the correct place. Let me know if we need more tests.

**Comment by asmeurer** [comment]
Oh yeah I misremembered. It wasn't worth it to implement these things: https://github.com/sympy/sympy/pull/14085#issuecomment-364575879. I think the rule I specified in https://github.com/sympy/sympy/pull/16437#issuecomment-476334620 is the correct one. You can double check the Python docs.

**Comment by smichr** [comment]
> can double check the Python docs

I verified it at the command line and added tests for each of the cases disallowed. Minor mods to the main code since it only has to be dealt with in one place, now.

**Comment by smichr** [comment]
This is ready to go when tests pass.

**Pull request merged**